radix-engine-profiling/src/info_alloc/mod.rs

```diff
@@ -0,0 +1,251 @@
+use std::{
+    alloc::{GlobalAlloc, Layout, System},
+    sync::atomic::{AtomicBool, AtomicIsize, Ordering},
+};
+
+#[global_allocator]
+pub static INFO_ALLOC: InfoAlloc<System> = InfoAlloc::new(System);
+
+thread_local! {
+    static INFO_ALLOC_DATA_TLS: InfoAllocData = InfoAllocData::new();
+}
+
+/// Heap allocations tracker
+///
+/// This allocator information provider can count allocations up to isize::MAX (9_223_372_036_854_775_807),
+/// in case of exceeding this value code will panic with message: 'Value out of range'.
+pub struct InfoAlloc<T: GlobalAlloc> {
+    /// Heap allocator to use, default usage: System
+    allocator: T,
+    /// Determine if allocation data gathering is enabled
+    enabled: AtomicBool,
+}
+
+/// Allocation data stored in Thread Local Store (separate data for each thread)
+pub struct InfoAllocData {
+    /// Sum of bytes allocated during measurements (for reallocations only additional memory is counted, no dealocation is counted)
+    sum_counter: AtomicIsize,
+    /// Current level of allocated bytes (allocation and deallocation are counted, incl. reallocation)
+    current_level: AtomicIsize,
+    /// Maximum level (peak) of allocated bytes (the max of this field and current_level)
+    max_level: AtomicIsize,
+}
+
+impl InfoAllocData {
+    fn new() -> Self {
+        Self {
+            sum_counter: AtomicIsize::new(0),
+            current_level: AtomicIsize::new(0),
+            max_level: AtomicIsize::new(0),
+        }
+    }
+}
+
+impl<T: GlobalAlloc> InfoAlloc<T> {
+    pub const fn new(allocator: T) -> Self {
+        InfoAlloc {
+            allocator,
+            enabled: AtomicBool::new(false),
+        }
+    }
+
+    pub fn set_enable(&self, enable: bool) {
+        self.enabled.store(enable, Ordering::Release);
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        self.enabled.load(Ordering::Acquire)
+    }
+
+    /// Resets internal counters. Usually used to start measurement.
+    pub fn reset_counters(&self) {
+        INFO_ALLOC_DATA_TLS.with(|val| {
+            val.sum_counter.store(0, Ordering::Release);
+            val.current_level.store(0, Ordering::Release);
+            val.max_level.store(0, Ordering::Release);
+        });
```

Comment:
Why do we need atomics if they're thread-local?

Reply:
Yes, this is a leftover from a previous implementation (when counter was shared between threads), should be fine to change to ordinary number type.

Reply:
I've changed counter types, but it requires usage of  `ReffCell` as thread local data due to needed interior mutability (atomics are implemented using `UnsafeCell` internally).